refactor: Split per-peer parts of net module into new node/connection module

ajtowns opened the pull request:
Move per-peer code from net into a new node/connection module. This is intended to capture logic that is lower-level than the p2p protocol (done in net_processing), but higher-level than raw bits-over-the-wire (handled by Sock).

ajtowns committed:
qt/guiutil: add missing include

ajtowns committed:
remove unnecessary net.h includes

ajtowns commented:
> `node/node` seems a bit confusing, because the two `node` refer to different things. `node` refers to this node and the other `node` refers to a peer, so `node/peer` may be better?

I did think about that, but it's confusing in its own way: we have `class Peer` in net_processing that contains the per-peer protocol-level data for a particular connection. I guess we could decide to move towards calling them `Peer` (instead of `CNode`) and `PeerLogic` (instead of `Peer`), and `PeerLogicManager` (instead of `PeerManager`)?

dergoegge commented:
While we are bike-shedding the naming, I actually think `node/connection` would make sense.  `Connection` would also seem like the appropriate name for `CNode` given that it is managed by something called connection manager.

kevkevinpal commented on src/test/validation_tests.cpp:
```diff
@@ -4,7 +4,6 @@
 
 #include <chainparams.h>
 #include <consensus/amount.h>
-#include <net.h>
```
was this header not needed? I don't see `#include <node/connection.h>` being added

ajtowns replied:
Seems like it wasn't; iwyu doesn't recommend adding it back, either. Shouldn't be needed -- validation stuff should be "kernel" level, so shouldn't rely on any p2p/net things.

kevkevinpal replied:
sounds good to me!

dergoegge commented on src/node/connection.h:
```diff
@@ -0,0 +1,887 @@
+    virtual bool ShouldReconnectV1() const noexcept = 0;
+};
+
+class V1Transport final : public Transport
```
What do you think of moving the transport impls to net.cpp (or their own module)?. They aren't really part of this interface, more of a implementation detail.

ajtowns replied:
The class info is exposed for unit tests, and you at least need `class Transport` in order to know what a `CNode` looks like. Separating out the V1/V2 implementation details alone doesn't seem like much of a win. (And wouldn't really be a win at all if we just had a concrete Transport class instead of the replaceable virtual setup, cf https://github.com/bitcoin/bitcoin/pull/28331#discussion_r1336030845)

ajtowns replied:
(The `Transport` stuff is all specific to a single connection, so it would go in `node/connection.cpp` rather than `net.cpp` per the logic of this PR)

ajtowns committed:
move-only: split CNode into new node/connection.h module

Move per-peer code from net into a new node/connection module. This
is intended to capture logic that is lower-level than the p2p protocol
(done in net_processing), but higher-level than raw bits-over-the-wire
(handled by Sock).

naumenkogs commented:
Concept ACK. Good thing this separation doesn't require additional wrappers and fields.

naumenkogs commented:
ACK fb03ca14c8076bc5fd3c9535c280996032db9cf6

ajtowns commented:
Closing as up-for-grabs (will grab it myself in a while if no one else does first)
